libbeat/processors/extract_array/extract_array.go

```diff
@@ -0,0 +1,156 @@
+			checks.AllowedFields("field", "mappings", "ignore_missing", "overwrite_keys", "fail_on_error", "when")))
+}
+
+// NewExtractArray builds a new extract_array processor.
+func NewExtractArray(c *common.Config) (processors.Processor, error) {
+	config := defaultExtractArrayConfig
+
+	err := c.Unpack(&config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unpack the extract_array configuration: %s", err)
+	}
+
+	if len(config.Mappings) == 0 {
+		return nil, errNoMappings
+	}
+
+	p := &extractArrayProcessor{
+		config:   config,
+		mappings: make([]fieldMapping, 0, len(config.Mappings)),
+	}
+	for field, column := range config.Mappings.Flatten() {
+		colIdx, ok := common.TryToInt(column)
+		if !ok || colIdx < 0 {
+			return nil, fmt.Errorf("bad extract_array mapping for field %s: %+v is not a positive integer", field, column)
+		}
+		p.mappings = append(p.mappings, fieldMapping{from: colIdx, to: field})
+	}
+	sort.Slice(p.mappings, func(i, j int) bool {
+		return p.mappings[i].from < p.mappings[j].from
+	})
+	return p, nil
+}
+
+func (f *extractArrayProcessor) Run(event *beat.Event) (*beat.Event, error) {
+	iValue, err := event.GetValue(f.config.Field)
+	if err != nil {
+		if f.config.IgnoreMissing && errors.Cause(err) == common.ErrKeyNotFound {
+			return event, nil
+		}
+		return event, errors.Wrapf(err, "could not fetch value for field %s", f.config.Field)
+	}
+
+	v := reflect.ValueOf(iValue)
+	if t := v.Type(); t.Kind() != reflect.Slice {
+		if !f.config.FailOnError {
+			return event, nil
+		}
+		return event, errors.Wrapf(err, "unsupported type for field %s: got: %s needed: array", f.config.Field, t.String())
+	}
+
+	saved := *event
+	if f.config.FailOnError {
+		saved.Fields = event.Fields.Clone()
+		saved.Meta = event.Meta.Clone()
+	}
+
+	n := v.Len()
+	for _, mapping := range f.mappings {
+		if mapping.from >= n {
+			if !f.config.FailOnError {
+				continue
+			}
+			return &saved, errors.Errorf("index %d exceeds length of %d when processing mapping for field %s", mapping.from, n, mapping.to)
+		}
+		if !f.config.OverwriteKeys {
+			if _, err = event.GetValue(mapping.to); err == nil {
+				if !f.config.FailOnError {
+					continue
+				}
+				return &saved, errors.Errorf("target field %s already has a value. Set the overwrite_keys flag or drop/rename the field first", mapping.to)
+			}
+		}
+		if _, err = event.PutValue(mapping.to, clone(v.Index(mapping.from).Interface())); err != nil {
+			if !f.config.FailOnError {
+				continue
+			}
+			return &saved, errors.Wrapf(err, "failed setting field %s", mapping.to)
+		}
+	}
+	return event, nil
+}
+
+func (f *extractArrayProcessor) String() (r string) {
+	return fmt.Sprintf("extract_array={field:%s, mappings:%v}", f.config.Field, f.mappings)
+}
+
+func clone(value interface{}) interface{} {
+	// TODO: This is dangerous but done by most processors.
+	//       Otherwise need to reflect value and deep copy lists / map types.
+	switch v := value.(type) {
+	case common.MapStr:
```

Comment:
Is is possible after going through the event normalization for one of the values to be an `map[string]interface{}`? Perhaps just to be defensive you should add a case statement for it and cast it to `common.MapStr` and then do the shallow `Clone()`.